### PR TITLE
Remove expandDefinitions infrastructure in the preprocessor

### PR DESCRIPTION
**Pull request by ajreynol**
Users of the method `Preprocess::expandDefinitions` included several more advanced features (e.g. abduction, interpolation) which should *not* provide input to subsolvers that involves internally introduced symbols (which ExpandDefinitions::expandDefinitions does). This corrects the issue.

This simplifies the functionality of the preprocessor to apply substitutions only.

Expand definitions is still required for `getValue`, where it is now used as a local utility.

### CODE REVIEW

**Comment by ajreynol** [comment]
I need to double check that `getValue` remains the same in this PR.

File: src/smt/preprocessor.h [hunk 1]
```diff
@@ -89,17 +88,17 @@ class Preprocessor : protected EnvObj
    */
   Node simplify(const Node& n);
   /**
-   * Expand the definitions in a term or formula n.  No other
-   * simplification or normalization is done.
+   * Apply top-level substitutions and eliminate abstract values in a term or
+   * formula n,  No other simplification or normalization is done.
```
Comment (aniemetz) [inline src/smt/preprocessor.h:92]:
```suggestion
   * formula n.  No other simplification or normalization is done.
```

**Review by aniemetz — APPROVED** [review]
One minor thing, else LGTM.

**Comment by ajreynol** [comment]
I double checked, `get-value` is actually still a reason to keep `expandDefinitions`.

I've refactored the PR so that expandDefinitions is a utility, but not part of the core preprocessor. The same fixes apply (to only use substituted form).

File: src/smt/solver_engine.cpp [hunk 2]
```diff
@@ -953,15 +954,24 @@ Node SolverEngine::simplify(const Node& ex)
...
+
+  // We must expand definitions here, which replaces certain subterms of t
+  // by the form that is used internally. This is necessary for some corner
+  // cases of get-value to be accurate, e.g. when getting the value of
+  // a division-by-zero term, we require getting the appropriate skolem
+  // function corresponding to division-by-zero which may have been used during
+  // the previous satisfiability check.
+  std::unordered_map<Node, Node> cache;
+  ExpandDefs expDef(*d_env.get());
+  // must apply substitutions first to ensure we expand definitions in the
```
Comment (aniemetz) [inline src/smt/solver_engine.cpp:971]:
```suggestion
  // Must apply substitutions first to ensure we expand definitions in the
```

File: src/smt/expand_definitions.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+namespace cvc5::internal {
+namespace smt {
+
+/**
+ * Implements expand definitions, which returns the expanded form of a term.
+ *
+ * This method is similar in nature to PropEngine::preprocess in that it
+ * converts a (possibly user-provided) term into the form that we pass
+ * internally. However, this method can be seen as a lightweight version
+ * of that method which only does enough conversions to make e.g. get-value
```
Comment (aniemetz) [inline src/smt/expand_definitions.h:35]:
```suggestion
 * of that method which only does enough conversions to make, e.g., get-value
```

File: src/smt/solver_engine.cpp [hunk 4]
```diff
@@ -953,15 +954,24 @@ Node SolverEngine::simplify(const Node& ex)
...
-  Node n = d_smtSolver->getPreprocessor()->applySubstitutions(ex);
+Node SolverEngine::getValue(const Node& t) const
+{
+  ensureWellFormedTerm(t, "get value");
+  Trace("smt") << "SMT getValue(" << t << ")" << endl;
+  TypeNode expectedType = t.getType();
+
+  // We must expand definitions here, which replaces certain subterms of t
+  // by the form that is used internally. This is necessary for some corner
+  // cases of get-value to be accurate, e.g. when getting the value of
```
Comment (aniemetz) [inline src/smt/solver_engine.cpp:965]:
```suggestion
  // cases of get-value to be accurate, e.g., when getting the value of
```

File: src/smt/expand_definitions.cpp [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+      else
+      {
+        node = n;
+      }
+      // the partial functions can fall through, in which case we still
+      // consider their children
+      worklist.push(std::make_tuple(
+          Node(n), node, true));  // Original and rewritten result
+
+      for (size_t i = 0; i < node.getNumChildren(); ++i)
```
Comment (aniemetz) [inline src/smt/expand_definitions.cpp:101]:
```suggestion
      for (size_t i = 0, nchild= node.getNumChildren(); i < nchild; ++i)
```

**Review by aniemetz — APPROVED** [review]

**Pull request merged**